File: mlp-ga-driving-java/src/driving/brain/Neuron.java
package driving.brain;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Yzumi on 27/12/2017.
 */
public class Neuron {

    private final List<Float> weights;
    private final float bias;
    private float output;

    public Neuron(final int numberOfconnections) {
        this.weights = new ArrayList<>();
        for(int i = 0; i < numberOfconnections; i++) {
            this.weights.add(this.generateRandomWeight());
        }
        this.bias = this.generateRandomWeight();
    }

    public Neuron(final List<Float> weights, final float bias) {
        this.weights = weights;
        this.bias = bias;
    }

    private float generateRandomWeight() {
        return (float) Math.random();
    }

    public void calculateOutput(final List<Float> inputs) {
        final float weightedInput = this.calculateWeightedInput(inputs) + this.bias;
        this.output = 1 / (1 + (float)Math.exp(-weightedInput));
    }

    private float calculateWeightedInput(final List<Float> inputs) {
        float ret = 0;
        for(int i = 0; i < inputs.size(); i++) {
            ret += this.weights.get(i) * inputs.get(i);
        }
        return ret;
    }

    public float getOutput() {
        return output;
    }

    public static Neuron reproduce(final Neuron neuron1, final Neuron neuron2, final float mutationRate) {
        final List<Float> childWeights = new ArrayList<>();
        for(int i = 0; i < neuron1.weights.size(); i++) {
            float childWeight;
            if(Math.random() < 0.5) {
                childWeight = neuron1.weights.get(i);
            } else {
                childWeight = neuron2.weights.get(i);
            }
            childWeight = Neuron.randomDeltaMutation(childWeight, mutationRate);
            childWeights.add(childWeight);
        }
        float bias;
        if(Math.random() < 0.5) {
            bias = neuron1.bias;
        } else {
            bias = neuron2.bias;
        }
        bias = Neuron.randomDeltaMutation(bias, mutationRate);
        return new Neuron(childWeights, bias);
    }

    private static float randomDeltaMutation(final float value, final float mutationRate) {
        final float deltaValue;
        if(Math.random() < mutationRate) {
            deltaValue = 2 * (float) Math.random() - 1;
        } else {
            deltaValue = 0;
        }
        return value + deltaValue;
    }

}

File: javascript/mlp/scripts/Neuron.js
class Neuron {

    constructor(numberOfConnections) {
        this.weights = [];
        for(var i = 0; i < numberOfConnections; i++) {
            this.weights.push(Math.random());
        }
        this.bias = Math.random();
        this.deltaOutput = 0;
        this.output = 0;
    }

    calculateOutput(inputs) {
        var weightedInput = this.calculateWeightedInput(inputs);
        weightedInput += this.bias;
        this.output = 1 / (1 + Math.exp(-weightedInput));
    }

    calculateWeightedInput(inputs) {
        var ret = 0;
        for(var i = 0; i < inputs.length; i++) {
            ret += inputs[i] * this.weights[i];
        }
        return ret;
    }

    adjustWeight(deltaWeight, i) {
        this.weights[i] += deltaWeight;
    }

    getOutput() {
        return this.output;
    }

    getDeltaOutput() {
        return this.deltaOutput;
    }

    setDeltaOutput(deltaOutput) {
        this.deltaOutput = deltaOutput;
    }

    getWeights() {
        return this.weights;
    }

    getWeight(i) {
        return this.weights[i];   
    }

    getBias() {
        return this.bias;
    }

    adjustBias(ijDeltaWeight) {
        this.bias += ijDeltaWeight;
    }

    print() {
        console.log("weights = " + this.weights);
        console.log("deltaOutput = " + this.deltaOutput);
        console.log("output = " + this.output);
        console.log("bias = " + this.bias);
    }

}